### PR TITLE
Various minor fixes to ada::url_aggregator

### PR DESCRIPTION
**Pull request by lemire**
1. adds `has_port`
2. speeds up ada::helpers::substring
3. guard `update_base_username` and `update_base_password`
4. fix silly typo `url.set_host(url.get_host());`.
5. fix `url_aggregator::parse_ipv4` in the case where we have a trailing dot.
6. simplifies `get_password()` and `get_username()`.

### CODE REVIEW

File: src/parser.cpp [hunk 1]
```diff
@@ -322,14 +322,18 @@ namespace ada::parser {
...
-              // and not having a password.
-              url.update_base_password(base_url->get_password());
-              url.set_hostname(base_url->get_hostname());
-  	          url.update_base_port(base_url->retrieve_base_port());
+              if(base_url->has_non_empty_username()) {
+                url.update_base_username(base_url->get_username());
+              }
+              if(base_url->has_non_empty_password()) {
+                url.update_base_password(base_url->get_password());
+              }
```
Comment (anonrig) [inline src/parser.cpp:330]: I think we should merge these two, like in my pull request. Faster, and a lot easier to handle the case of `//`

File: src/parser.cpp [hunk 2]
```diff
@@ -322,14 +322,18 @@ namespace ada::parser {
...
+              if(base_url->has_authority()) {
+                url.set_hostname(base_url->get_hostname());
+                if(base_url->has_port())  {
+                  url.update_base_port(base_url->retrieve_base_port());
+                }
```
Comment (anonrig) [inline src/parser.cpp:335]: This is wrong, because if it's not existing, then we should set it as omitted?

File: src/url_aggregator.cpp [hunk 3]
```diff
@@ -426,6 +426,7 @@ bool url_aggregator::set_host_or_hostname(const std::string_view input) {
     // Let host be the result of host parsing host_view with url is not special.
     if (host_view.empty()) {
       clear_base_hostname();
+      add_authority_slashes_if_needed();
```
Comment (anonrig) [inline src/url_aggregator.cpp:429]: should we just move this line to `clear_base_hostname()` function?

File: src/parser.cpp (hunk 1 above)
Comment (lemire) [reply src/parser.cpp:330]: Merging them is faster, but your PR creates temporary std::strings...

File: src/url_aggregator.cpp (hunk 3 above)
Comment (lemire) [reply src/url_aggregator.cpp:429]: Let us try that.

File: src/parser.cpp (hunk 2 above)
Comment (lemire) [reply src/parser.cpp:335]: Ok. Let me pull this out.

File: src/parser.cpp (hunk 1 above)
Comment (anonrig) [reply src/parser.cpp:330]: Yes, I intend to remove it before merging. Its wip at the moment

**Review by anonrig — APPROVED** [review]

**Pull request merged**